expo/expo
# fix(cli): use new `@react-native/dev-middleware` for debugging

# Why

This PR re-implements the debugging tools, temporarily removed to merge the React Native `0.73.0-rc.x` upgrade.

In this re-implementation, we are also trying to move over to the new React Native JS Inspector, the one shipped with React Native 0.73.x.

![image](https://github.com/expo/expo/assets/1203991/d747a6cc-9c89-4279-8d3a-496fbb5a7cce)

# TODO

There are still pending/follow-up tasks to complete the full debugging transition.

- [ ] (Optional) try to get rid of the last proxy workarounds, besides using different `Device` class
- [ ] Fix `@react-native/dev-middleware` not responding to `/open-debugger` (request times out, but works)
- [ ] Enable network inspector inside Chrome DevTools from `@react-native/dev-middleware`
- [ ] Use file tree instead of snippet workspaces, under `sources` tab in Chrome DevTools from `@react-native/dev-middleware`
- [ ] Update `vscode-expo-tools` to provide `?userAgent=`, since we don't get any `user-agent` header from the extension.

# How

- Refactored existing `InspectorProxy`, `InspectorDevice`, and all handles into new `debugging` group under `src/start/server/metro/debugging`.
- Integrated `@react-native/dev-middleware` into separate `createDebugMiddleware` function
- Extend `InspectorProxy` instead of wrapping it
- Cleaned up no-longer-necessary workarounds and overwrites in `InspectorProxy` and `InspectorDevice`

# Test Plan

Prepare the current template as npm tarball

- `$ cd .../expo/expo`
- `$ cd ./templates/expo-template-bare-minimum/`
- `$ npm pack`

Prepare a new canary project

- `$ bun create expo ./test-debug --template blank`
- `$ bun expo install expo@canary`
- `$ expod install --fix` (_`expod` → run `@expo/cli` from source_)
- `$ expod prebuild --template ./<tarball-from-above>.tgz`
- `$ expod run:ios`
- Open the _two_ debuggers:
  - **Press J in terminal** → Current Expo Chrome DevTools implementation
  - `$ curl -X POST "http://localhost:8081/open-debugger"` → Experimental Chrome DevTools
  - `$ curl -X POST "http://localhost:8081/open-debugger?appId=<bundleIdentifier>"` → Experimental Chrome DevTools
  - `$ curl -X POST "http://localhost:8081/open-debugger?device=<deviceIdentifier>"` → Experimental Chrome DevTools (see [device ID for android](https://github.com/facebook/react-native/blob/c0375b8dadaafc0e0c11bbe676a61c44ac9fab1d/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java#L317C20-L317C40))

# Checklist
